R/module_teal_with_splash.R

```diff
@@ -52,36 +52,59 @@ ui_teal_with_splash <- function(id,
 #'   will be displayed in the teal application. See [modules()] and [module()] for
 #'   more details.
 #' @inheritParams shiny::moduleServer
-#' @return `reactive`, return value of [srv_teal()]
+#' @return `reactive` containing `teal_data` object when data is loaded.
+#' If data is not loaded yet, `reactive` returns `NULL`.
 #' @export
 srv_teal_with_splash <- function(id, data, modules, filter = teal_slices()) {
-  checkmate::assert_class(data, "TealDataAbstract")
+  checkmate::assert_multi_class(data, c("TealDataAbstract", "teal_data"))
   moduleServer(id, function(input, output, session) {
-    logger::log_trace(
-      "srv_teal_with_splash initializing module with data { paste(data$get_datanames(), collapse = ' ')}."
-    )
+    logger::log_trace("srv_teal_with_splash initializing module with data { toString(get_dataname(data))}.")
 
     if (getOption("teal.show_js_log", default = FALSE)) {
       shinyjs::showLog()
     }
 
-    is_pulled_data <- teal.data::is_pulled(data)
     # raw_data contains TealDataAbstract, i.e. R6 object and container for data
     # reactive to get data through delayed loading
     # we must leave it inside the server because of callModule which needs to pick up the right session
-    if (is_pulled_data) {
-      raw_data <- reactiveVal(data) # will trigger by setting it
+    raw_data <- if (inherits(data, "teal_data")) {
+      reactiveVal(data)
+    } else if (teal.data::is_pulled(data)) {
+      new_data <- do.call(
+        teal.data::teal_data,
+        c(
+          lapply(data$get_datasets(), function(x) x$get_raw_data()),
+          code = data$get_code(),
+          join_keys = data$get_join_keys()
+        )
+      )
+      reactiveVal(new_data) # will trigger by setting it
     } else {
-      raw_data <- data$get_server()(id = "startapp_module")
+      raw_data_old <- data$get_server()(id = "startapp_module")
+      raw_data <- reactive({
+        data <- raw_data_old()
+        if (!is.null(data)) {
+          # raw_data is a reactive which returns data only when submit button clicked
+          # otherwise it returns NULL
+          do.call(
+            teal.data::teal_data,
+            c(
+              lapply(data$get_datasets(), function(x) x$get_raw_data()),
+              code = data$get_code(),
+              join_keys = data$get_join_keys()
+            )
+          )
```

Comment:
This is duplicated. Can it be avoided?

Reply:
will be removed soon also. Conversion from TealData to teal_data is temporary